### PR TITLE
draft gs-compatible pipeline added

### PR DESCRIPTION
**Pull request by evanmauceli**

### CODE REVIEW

File: demuliplex.sh [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+# Environment setup
+################################################################################
+P=`dirname $0`
+source $P/common.sh
+initialize
+
+num_samples=`wc -l $SAMPLEFILE | awk '{print $1 -1}'`
+num_lanes=`grep LaneCount $RUN/RunInfo.xml | awk '{print $2}' | awk -F"=" '{print $2}' | sed s/\"/""/g`
+flowcell=`grep -m 1 Flowcell $RUN/RunInfo.xml | sed "s/.*<Flowcell>\(.*\)<\/Flowcell>.*/\1/"`
+num_barcodes=`echo $READSTRUCT | sed 's/\(.\)/&\n/g' | grep B | wc -l`
```
Comment (tfenne) [inline demuliplex.sh:92]: A (slightly) nicer way to do this:
```
num_barcodes=$(echo $READSTRUCT | tr -cd "B" | wc -c)
```
Uses `tr` to delete any non-`B` characters and then counts the remaining characters.

File: gs-pipeline.sh [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            "--output $out_dir/tmp.germline.unfiltered.vcf.gz"
+
+    execute "java -Xmx4096m -jar $picard FilterVcf" \
+             "VALIDATION_STRINGENCY=SILENT" \
+             "CREATE_INDEX=true" \
+             "I=$out_dir/tmp.germline.unfiltered.vcf.gz" \
+             "O=$germline_vcf" \
+             "MIN_AB=0.2 MIN_DP=0 MIN_GQ=20 MAX_FS=50.0 MIN_QD=6.0"
+
+    #execute "rm $out_dir/tmp.g.vcf.gz $out_dir/tmp.germline.unfiltered.vcf.gz"
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply gs-pipeline.sh:200]: removed

File: demuliplex.sh [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+################################################################################
+# Environment setup
+################################################################################
+P=`dirname $0`
+source $P/common.sh
+initialize
+
+num_samples=`wc -l $SAMPLEFILE | awk '{print $1 -1}'`
+num_lanes=`grep LaneCount $RUN/RunInfo.xml | awk '{print $2}' | awk -F"=" '{print $2}' | sed s/\"/""/g`
+flowcell=`grep -m 1 Flowcell $RUN/RunInfo.xml | sed "s/.*<Flowcell>\(.*\)<\/Flowcell>.*/\1/"`
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:91]: after much flailing, changed as suggested

File: demuliplex.sh [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+
+################################################################################
+# Environment setup
+################################################################################
+P=`dirname $0`
+source $P/common.sh
+initialize
+
+num_samples=`wc -l $SAMPLEFILE | awk '{print $1 -1}'`
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:89]: changed as suggested

File: demuliplex.sh [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+if [ $num_barcodes -eq 1 ]; then
+  header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT_PREFIX","SAMPLE_ALIAS","LIBRARY_NAME","BARCODE_1"}'`
+  extr_header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT","SAMPLE_ALIAS","LIBRARY_NAME","barcode_sequence_1"}'`
+elif [ $num_barcodes -eq 2 ]; then
+  header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT_PREFIX","SAMPLE_ALIAS","LIBRARY_NAME","BARCODE_1","BARCODE_2"}'`
+  extr_header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT","SAMPLE_ALIAS","LIBRARY_NAME","barcode_sequence_1","barcode_sequence_2"}'`
+else
+  header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT_PREFIX","SAMPLE_ALIAS","LIBRARY_NAME"}'`
+  extr_header=`awk -v OFS='\t' 'BEGIN {print "OUTPUT","SAMPLE_ALIAS","LIBRARY_NAME"}'`
+fi
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:106]: changed as suggested

File: demuliplex.sh [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                "OUTPUT_DIR=$RUN/Data/Intensities/BaseCalls/ READ_STRUCTURE=$READSTRUCT BARCODE_FILE=$lib_extract_file" \
+                "METRICS_FILE=$RUN/Data/Intensities/BaseCalls/barcode_counts.lane-$lane.metrics.txt" \
+                "COMPRESS_OUTPUTS=true NUM_PROCESSORS=4 MAX_MISMATCHES=2 LANE=$lane"
+    fi
+
+    execute "java -Xmx12g -jar $picard IlluminaBasecallsToFastq" \
+            "VALIDATION_STRINGENCY=SILENT CREATE_INDEX=true BASECALLS_DIR=$RUN/Data/Intensities/BaseCalls/" \
+            "READ_STRUCTURE=$READSTRUCT MULTIPLEX_PARAMS=$lib_params_file" \
+            "INCLUDE_NON_PF_READS=true IGNORE_UNEXPECTED_BARCODES=true LANE=$lane RUN_BARCODE=$flowcell" \
+            "NUM_PROCESSORS=4 READ_NAME_FORMAT=ILLUMINA"
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:151]: 1. a good question!  It works for IlluminaBasecallsToSam, but will make sure to test on IlluminaBasecallsToFastq
2. Yikes!  I don't even think this is an arg for IlluminaBasecallsToFastq. Will fix copy/paste errors throughout this call.
3. Nope
4. changed as suggested

File: demuliplex.sh [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    echo "$header" > $lib_params_file
+    echo "$extr_header" > $lib_extract_file
+
+    {
+    read
+    while read -r sample barcode1 barcode2;
+    do
+        if [ $num_barcodes -eq 1 ]; then
+            awk -v var=$lane -v out="$OUT" -v s="$sample" -v b1="$barcode1" -v b2="$barcode2" \
+                'BEGIN {print out"/"s".l"var"\t"s"\t"s"\t"b1}' >> $lib_params_file
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:124]: modified as suggested

File: demuliplex.sh (hunk 1 above)
Comment (evanmauceli) [reply demuliplex.sh:92]: changed as suggested

File: gs-pipeline.sh [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+
+################################################################################
+# Run the pipeline
+################################################################################
+
+if [ ! -f $P/bin/gatk.jar ]; then
+    banner "Downloading GATK4..."
+    if $(which wget > /dev/null); then
+         wget -qO $P/gatk.zip $GATK_URL
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply gs-pipeline.sh:106]: removed extra space in both places.  and I did notice that the python person called this out

File: demuliplex.sh [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+if [[ -z "$RUN" ]]; then usage; echo; echo "Error: run directory parameter is required."; exit 1; fi
+if [[ -z "$OUT" ]]; then usage; echo; echo "Error: output directory parameter is required."; exit 1; fi
+if [[ -z "$SAMPLEFILE" ]]; then usage; echo; echo "Error: sample file parameter is required."; exit 1; fi
+if [[ -z "$READSTRUCT" ]]; then usage; echo; echo "Error: read structure parameter is required."; exit 1; fi
+
+
+################################################################################
+# Environment setup
+################################################################################
+P=`dirname $0`
```
[the remark this answers is not in the document]
Comment (evanmauceli) [reply demuliplex.sh:85]: replaced backpacks with $() notation throughout the pipelines

**Pull request merged**